src/main/java/org/solid/testharness/http/Client.java

```diff
@@ -210,38 +177,37 @@ public String getUser() {
         } else {
             headers.put(HttpConstants.HEADER_AUTHORIZATION, HttpConstants.PREFIX_BEARER + accessToken);
         }
-        if (agent != null) {
-            headers.put(HttpConstants.USER_AGENT, agent);
-        }
+        headers.put(HttpConstants.USER_AGENT, agent);
         return headers;
     }
 
-    // TODO: Switch to elliptical curve as it is faster
-    public String generateDpopToken(final String htm, final String htu) {
-        if (clientKey == null) {
-            throw new RuntimeException("This instance does not have DPoP support added");
-        }
-        try {
-            final JsonWebSignature jws = new JsonWebSignature();
-//            jws.setAlgorithmHeaderValue(AlgorithmIdentifiers.ECDSA_USING_P256_CURVE_AND_SHA256);
-            jws.setAlgorithmHeaderValue(AlgorithmIdentifiers.RSA_USING_SHA256);
-            jws.setHeader(TYPE, "dpop+jwt");
-            jws.setJwkHeader(clientKey);
-            jws.setKey(clientKey.getPrivateKey());
-
-            final JwtClaims claims = new JwtClaims();
-            claims.setJwtId(randomUUID().toString());
-            claims.setStringClaim("htm", htm);
-            claims.setStringClaim("htu", htu);
-            claims.setIssuedAtToNow();
-
-            jws.setPayload(claims.toJson());
-
-            return jws.getCompactSerialization();
-        } catch (final JoseException ex) {
-            throw new UncheckedJoseException("Unable to generate DPoP token", ex);
+    private HttpRequest.Builder authorize(@NotNull final HttpRequest.Builder builder) {
+        requireNonNull(builder, "builder is required");
+        if (accessToken == null) return builder;
+        if (dpopSupported) {
+            builder.setHeader(HttpConstants.HEADER_AUTHORIZATION, HttpConstants.PREFIX_DPOP + accessToken);
+            return signRequest(builder);
+        } else {
+            return builder.setHeader(HttpConstants.HEADER_AUTHORIZATION, HttpConstants.PREFIX_BEARER + accessToken);
         }
     }
 
+    private String generateDpopToken(final String htm, final String htu) {
+        requireNonNull(clientKey, "This instance does not have DPoP support added");
+        final JwtClaims claims = new JwtClaims();
+        claims.setJwtId(randomUUID().toString());
+        claims.setStringClaim("htm", htm);
+        claims.setStringClaim("htu", htu);
+        claims.setIssuedAtToNow();
+        return JwsUtils.generateDpopToken(clientKey, claims);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("Client: user=%s, dPoP=%s, session=%s, local=%s",
+                user, dpopSupported, httpClient.cookieHandler().isPresent(),
+                System.getProperty("jdk.internal.httpclient.disableHostnameVerification", "false"));
+    }
+
     private Client() { }
```

Comment:
A response from a different channel:
When this project started there was no Solid Java library - work is starting on it soon and I agree the ecosystem needs core aspects of Solid abstracting in this way. However, in a test harness I wonder if there is a benefit to staying with lower level libraries. Some of the bugs I have uncovered so far in NSS and CSS are because I wasn't using libraries that have built in assumptions tied to supporting the use cases in the existing Solid ecosystem.

Reply:
There is already a task defined to look at using a Mock IdP